apis/common/v1/resource.go

```diff
@@ -77,10 +77,47 @@ type SecretKeySelector struct {
 	Key string `json:"key"`
 }
 
+// A ReferenceResolutionPolicy is a value for resolution policy for reference.
+type ReferenceResolutionPolicy string
+
+const (
+	// ReferencePolicyOptional is a resolution option.
+	// When the ReferenceResolutionPolicy is set to ReferencePolicyOptional the
+	// execution could continue even if the reference cannot be resolved.
+	ReferencePolicyOptional ReferenceResolutionPolicy = "Optional"
+
+	// ReferencePolicyAlways is a resolution option.
+	// When the ReferenceResolutionPolicy is set to ReferencePolicyAlways the
+	// reference will be tried to resolve for every reconcile loop.
+	ReferencePolicyAlways ReferenceResolutionPolicy = "Always"
+)
+
 // A Reference to a named object.
 type Reference struct {
 	// Name of the referenced object.
 	Name string `json:"name"`
+	// Policies of the referenced object.
+	Policies []ReferenceResolutionPolicy `json:"policies,omitempty"`
```

Comment:
I'm not sure if the array is the best way to represent that configuration since it can take arbitrary number of policies whereas we only have two knobs to configure. Two distinct fields with policy string values seem like the way to go (see last point [here](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#primitive-types)) but I'm struggling with coming up with names for those fields. What do you think about something like the following?
```yaml
vpcIdRef:
  name: my-vpc
  policy:
    resolve: Always # "IfNotPresent" by default.
    require: Never # "IfNotPresent" by default. "Always" could be the third option in the future.
```

Reply:
This is a tricky one - I'm also having a tough time with naming. Writing it all out, I think we're trying to configure:

1. When the reference is resolved. Either only when the corresponding field is unset ("not present"), or always.
2. Whether resolution of this reference is optional or required.

We have a similar case to 2 already in our [`PatchPolicy` type](https://github.com/negz/crossplane/blob/2ad8a10/apis/apiextensions/v1/composition_patches.go#L55), where we have (e.g.) `fromFieldPath: Optional`. With that in mind I'd make a small tweak:

```yaml
vpcIdRef:
  name: my-vpc
  policy:
    # Resolve specifies when this reference should be resolved. The default
    # is 'IfNotPresent', which will attempt to resolve the reference only when
    # the corresponding field is not present. Use 'Always' to resolve the
    # reference on every reconcile.
    resolve: Always
    # Resolution specifies whether resolution of this reference is required.
    # The default is 'Required', which means the reconcile will fail if the
    # reference cannot be resolved. 'Optional' means this reference will be
    # a no-op if it cannot be resolved.
    resolution: Optional
```

As a side-note, I'm still wary of putting the policy _inside_ the reference itself, but I can't think of a better approach (apart from fixing this in generic references instead). This used to be explicitly disallowed by API conventions but that restriction appears to have been removed.